eos-updater-flatpak-installer/main.c

```diff
@@ -0,0 +1,246 @@
+
+          if (!eos_updater_flatpak_installer_check_ref_actions_applied (installation,
+                                                                        pending_flatpak_deployments_state_path,
+                                                                        squashed_ref_actions_to_check,
+                                                                        &error))
+            {
+              g_message ("Flatpak ref actions are not up to date: %s",
+                         error->message);
+              return EXIT_CHECK_FAILED;
+            }
+
+          break;
+        }
+      case EU_INSTALLER_MODE_PERFORM:
+      case EU_INSTALLER_MODE_STAMP:
+        {
+          g_autoptr(GHashTable) new_flatpak_ref_actions_to_apply =
+            eos_updater_flatpak_installer_determine_flatpak_ref_actions_to_apply (NULL, &error);
+          g_autoptr(GPtrArray) squashed_ref_actions_to_apply =
+            eos_updater_util_flatten_flatpak_ref_actions_table (new_flatpak_ref_actions_to_apply);
+          g_autofree gchar *formatted_flatpak_ref_actions_to_apply = NULL;
+          g_autofree gchar *formatted_ordered_flatpak_ref_actions_to_apply = NULL;
+
+          if (new_flatpak_ref_actions_to_apply == NULL)
+            {
+              g_message ("Could not get information on which flatpak ref actions to check: %s",
+                         error->message);
+              return EXIT_FAILURE;
+            }
+
+          formatted_flatpak_ref_actions_to_apply =
+            eos_updater_util_format_all_flatpak_ref_actions ("All flatpak ref actions that are not yet applied",
+                                                             new_flatpak_ref_actions_to_apply);
+
+          g_message ("%s", formatted_flatpak_ref_actions_to_apply);
+
+          formatted_ordered_flatpak_ref_actions_to_apply =
+            eos_updater_util_format_flatpak_ref_actions_array ("Order in which actions will be applied",
+                                                               squashed_ref_actions_to_apply);
+          g_message ("%s", formatted_ordered_flatpak_ref_actions_to_apply);
+
+          if (!eos_updater_flatpak_installer_apply_flatpak_ref_actions (installation,
+                                                                        eos_updater_util_pending_flatpak_deployments_state_path (),
+                                                                        squashed_ref_actions_to_apply,
+                                                                        parsed_mode,
+                                                                        also_pull ? EU_INSTALLER_FLAGS_ALSO_PULL : EU_INSTALLER_FLAGS_NONE,
+                                                                        &error))
+            {
+               g_message ("Couldn't apply some flatpak update actions for this boot: %s",
+                          error->message);
+               return EXIT_FAILURE;
```

Comment:
Use `fail()` here.

Reply:
It would also be useful to have separate error codes for different types of failure. For example, see what `eos-updater-avahi.c` does (grep for `main() exit codes`).